hs_core/xmlparser.py

```diff
@@ -0,0 +1,61 @@
+import base64
+import os.path
+from StringIO import StringIO
+from xml.sax import xmlreader, saxutils
+from xml.sax.expatreader import ExpatParser
+
+from django.conf import settings
+
+
+XML_CACHE = os.path.join(settings.PROJECT_ROOT, '_xmlcache')
+
+
+class CatalogedExpatParser(ExpatParser):
+    def external_entity_ref(self, context, base, sysid, pubid):
+        if not self._external_ges:
+            return 1
+
+        source = self._ent_handler.resolveEntity(pubid, sysid)
+        source = saxutils.prepare_input_source(source,
+                                               self._source.getSystemId() or
+                                               "")
+
+        # If an entry does not exist in the xml cache, create it.
+        filepath = os.path.join(XML_CACHE, base64.urlsafe_b64encode(pubid))
+        if not os.path.isfile(filepath):
+            with open(filepath, 'w') as f:
+                contents = source.getByteStream().read()
+                source.setByteStream(StringIO(contents))
+                f.write(contents)
```

Comment:
@hyi It's on line 24, right at the very end.  I base64 encode the pubid to turn it into a safe file name (the real pubid has slashes and spaces, at the very least) and then open that file in write-mode if it doesn't exist.  These DTDs should be fairly stable (the one brought in by the test has a date as part of the pubid), so it should be ok to just leave the files sit.  Even if it turns out that we need a newer version of one, deleting the associated file can easily be done as a one-off action by a system administrator.